### ISSUE TITLE
Check tempfile.TemporaryFile() and siblings for unspecified-encoding (W1514)

### ISSUE THREAD

**Issue by stephen-dexda** [PyCQA/pylint#5290]
### Current problem

Currently `tempfile.TemporaryFile()` and siblings don't generate an `unspecified-encoding` warning when no explicit encoding is supplied. Presumably it should, in similar fashion to `open()`, `pathlib.Path().write_text()` etc.

### Desired solution

Check uses of the relevant `tempfile` classes for lack of explicit encodings.

**Comment by DanielNoord** [comment]
For future reference I have already created tests to add to `tests/functional/u/unspecified_encoding_py38.py`:

```python
# Tests for the tempfile module

tempfile.TemporaryFile()  # [unspecified-encoding]
tempfile.TemporaryFile("r")  # [unspecified-encoding]
tempfile.TemporaryFile("wb")
tempfile.TemporaryFile("w+b")
tempfile.TemporaryFile("wt")  # [unspecified-encoding]
tempfile.TemporaryFile("w+")  # [unspecified-encoding]
tempfile.TemporaryFile("w", encoding="utf-8")
tempfile.TemporaryFile("w", encoding=None)  # [unspecified-encoding]

LOCALE_ENCODING = locale.getlocale()[1]
tempfile.TemporaryFile("w", encoding=LOCALE_ENCODING)
with tempfile.TemporaryFile("w", encoding=LOCALE_ENCODING) as f:
    pass

LOCALE_ENCODING = None
tempfile.TemporaryFile("w", encoding=LOCALE_ENCODING)  # [unspecified-encoding]
with tempfile.TemporaryFile(  # [unspecified-encoding]
    "w", encoding=LOCALE_ENCODING
) as f:
    pass

with tempfile.TemporaryFile() as f:  # [unspecified-encoding]
    pass

tempfile.NamedTemporaryFile()  # [unspecified-encoding]
tempfile.NamedTemporaryFile("r")  # [unspecified-encoding]
tempfile.NamedTemporaryFile("wb")
tempfile.NamedTemporaryFile("w+b")
tempfile.NamedTemporaryFile("wt")  # [unspecified-encoding]
tempfile.NamedTemporaryFile("w+")  # [unspecified-encoding]
tempfile.NamedTemporaryFile("w", encoding="utf-8")
tempfile.NamedTemporaryFile("w", encoding=None)  # [unspecified-encoding]

LOCALE_ENCODING = locale.getlocale()[1]
tempfile.NamedTemporaryFile("w", encoding=LOCALE_ENCODING)
with tempfile.NamedTemporaryFile("w", encoding=LOCALE_ENCODING) as f:
    pass

LOCALE_ENCODING = None
tempfile.NamedTemporaryFile("w", encoding=LOCALE_ENCODING)  # [unspecified-encoding]
with tempfile.NamedTemporaryFile(  # [unspecified-encoding]
    "w", encoding=LOCALE_ENCODING
) as f:
    pass

with tempfile.NamedTemporaryFile() as f:  # [unspecified-encoding]
    pass

tempfile.SpooledTemporaryFile()  # [unspecified-encoding]
tempfile.SpooledTemporaryFile("r")  # [unspecified-encoding]
tempfile.SpooledTemporaryFile("wb")
tempfile.SpooledTemporaryFile("w+b")
tempfile.SpooledTemporaryFile("wt")  # [unspecified-encoding]
tempfile.SpooledTemporaryFile("w+")  # [unspecified-encoding]
tempfile.SpooledTemporaryFile("w", encoding="utf-8")
tempfile.SpooledTemporaryFile("w", encoding=None)  # [unspecified-encoding]

LOCALE_ENCODING = locale.getlocale()[1]
tempfile.SpooledTemporaryFile("w", encoding=LOCALE_ENCODING)
with tempfile.SpooledTemporaryFile("w", encoding=LOCALE_ENCODING) as f:
    pass

LOCALE_ENCODING = None
tempfile.SpooledTemporaryFile("w", encoding=LOCALE_ENCODING)  # [unspecified-encoding]
with tempfile.SpooledTemporaryFile(  # [unspecified-encoding]
    "w", encoding=LOCALE_ENCODING
) as f:
    pass

with tempfile.SpooledTemporaryFile() as f:  # [unspecified-encoding]
    pass
```

The problem with implementing this is that `TemporaryFile()` does not return a single inference option and we currently iterate over all inference results for this checker. Therefore we raise a warning for each result, effectively duplicating the warnings.
I did not immediately know how to solve this as we can't now for certain whether the first or second inference result is "better".

**Comment by Pierre-Sassoulas** [comment]
I upgraded the example, we still need to specify what we do when multiple inference are looped over..

**Comment by DanielNoord** [comment]
I have test cases written for this. I'm waiting for the `is None` constraints to implement `sys` constraints in astroid.

**Comment by prabhu** [comment]
One of my project broke due to unspecified-encoding false positive.

ValueError: binary mode doesn't take an encoding argument.

```python
def __init__(self, mode="wb", delete=True):
    self._mode = mode

# Later on
self._tempFile = open(file_name, self._mode, encoding="utf-8") # This should not be flagged.
```
